### PR TITLE
ft(S3C-3005): Ingest API

### PR DESCRIPTION
**Pull request by tmacro**

### CODE REVIEW

**Review by alexanderchan-scality — APPROVED** [review]

File: libV2/server/middleware.js [hunk 1]
```diff
@@ -39,9 +40,18 @@ function loggerMiddleware(req, res, next) {
 // next is purposely not called as all error responses are handled here
 // eslint-disable-next-line no-unused-vars
 function errorMiddleware(err, req, res, next) {
-    const code = err.code || 500;
+    let code = err.code || 500;
     let message = err.message || 'Internal Error';
 
+    // failed request validation by oas-tools
+    if (err.failedValidation) {
+        // You can't actually use destructing here
```
Comment (miniscruff) [inline libV2/server/middleware.js:48]: why not?

Comment (tmacro) [reply libV2/server/middleware.js:48]:
```
Tmac@Kennuc: /tmp cat test.js
let x;
let y;

{ x, y } = { x:1, y:2}
Tmac@Kennuc: /tmp node test.js
/tmp/test.js:4
{ x, y } = { x:1, y:2}
         ^

SyntaxError: Unexpected token '='
    at wrapSafe (internal/modules/cjs/loader.js:1054:16)
    at Module._compile (internal/modules/cjs/loader.js:1102:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47
```

Comment (miniscruff) [reply libV2/server/middleware.js:48]: hmm seems like it doesn't like destructuring onto existing objects... good to know. Not sure if the comment is required but yea.

Comment (miniscruff) [reply libV2/server/middleware.js:48]: Note that from the node REPL it works just fine which is confusing...
```shell
Welcome to Node.js v14.1.0.
Type ".help" for more information.
> let x, y;
undefined
> { x, y } = { x:1, y:2 }
{ x: 1, y: 2 }
> console.log(x)
1
undefined
> console.log(y)
2
undefined
```

File: tests/unit/v2/API/metrics/testIngestMetric.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+const events = generateFakeEvents(1, 50, 50);
+
+describe('Test ingestMetric', () => {
+    let ctx;
+    beforeEach(() => {
+        ctx = templateContext();
+    });
+
+    afterEach(() => sinon.restore());
```
Comment (miniscruff) [inline tests/unit/v2/API/metrics/testIngestMetric.js:16]: I think this also works
```suggestion
    afterEach(sinon.restore);
```

File: libV2/server/API/metrics/ingestMetric.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    try {
+        metrics = params.body.map(m => new UtapiMetric(m));
+    } catch (error) {
+        throw errors.InvalidRequest;
+    }
+    try {
+        await Promise.all(metrics.map(m => cacheClient.pushMetric(m)));
+    } catch (error) {
+        throw errors.ServiceUnavailable;
+    }
```
Comment (miniscruff) [inline libV2/server/API/metrics/ingestMetric.js:16]: You could combine the catches I think
```suggestion
    try {
        let metrics = params.body.map(m => new UtapiMetric(m));
        await Promise.all(metrics.map(m => cacheClient.pushMetric(m)));
    } catch (error) {
        throw errors.InvalidRequest || errors.ServiceUnavailable;
    }
```

Not sure if this any more readable but maybe.

**Review by miniscruff — APPROVED** [review]
Small ideas looks good otherwise

File: libV2/server/API/metrics/ingestMetric.js (hunk 3 above)
Comment (tmacro) [reply libV2/server/API/metrics/ingestMetric.js:16]: Wouldn't this `throw errors.InvalidRequest || errors.ServiceUnavailable;` always throw `errors.InvalidRequest`?

Comment (miniscruff) [reply libV2/server/API/metrics/ingestMetric.js:16]: Edit: I misread the original code oops... I thought it was `error.InvalidRequest` as in the error was raised which had an object as part of it. And not just grabbing an error from the errors object.

File: tests/unit/v2/API/metrics/testIngestMetric.js (hunk 2 above)
Comment (tmacro) [reply tests/unit/v2/API/metrics/testIngestMetric.js:16]: sinon looses some sort of internal state without the `() =>` and throws an error

Comment (miniscruff) [reply tests/unit/v2/API/metrics/testIngestMetric.js:16]: hmm weird, will have to remember this for later

**Pull request merged**